### ISSUE TITLE
XGB: Attention-CNN-LSTM + XGBoost Hybrid Forecasting

### ISSUE THREAD

**Issue by Gopi0318** [Rivier-Computer-Science/AI-Agent-Stock-Prediction#589]
**Description:**
Build a two-stage forecasting pipeline that:

Pretrains a 1D-CNN with self-attention feeding into an LSTM directly on raw OHLCV windows to learn temporal patterns.

Extracts the LSTM’s final hidden-state embeddings and refines them with an XGBoost model to predict next-day price movement or returns—all without any hand-crafted technical indicators.

**Acceptance Criteria:**

- Windowed Data: OHLCV is fetched, cleaned, normalized, and cut into fixed-length windows (train/val/test).
- Deep Model: The CNN+Attention+LSTM trains end-to-end, producing meaningful embeddings (no crashes, converges on validation).
- Embedding Export: Hidden-state vectors for every window are saved alongside true labels.
- XGBoost Tuning: XGBoost trains on those embeddings, improves hold-out metrics over the base deep model.
- JSON Output: Given a new window, the system quickly returns:
{ "date": "...", "cnn_lstm_score": 0.12, "xgb_prob": 0.78 }

Crew AI Decision: A Crew AI agent consumes that JSON and outputs exactly {"recommendation":"BUY"/"SELL"/"HOLD"}.

**Conditions of Satisfaction:**

- Accuracy: Model outputs match expected shapes and value ranges; embeddings dimension is correct.
- Speed: Full inference (window → CNN-LSTM → XGBoost → JSON) completes in ≤ 5 s on sample data.
- Robustness: Pipeline handles edge cases (e.g., missing bars) without errors.
- Integration: Crew AI agent consistently applies probability thresholds and returns valid recommendations.

**Definition of Done:**

- Data ingestion, cleaning, normalization, and windowing scripts are in place and tested.
- CNN+Attention+LSTM model code, training loop, and checkpointing are implemented.
- Embedding extraction utility correctly dumps hidden states with labels.
- XGBoost fine-tuning pipeline (hyperparameter search, training, evaluation) is complete.
- Inference script loads both models and produces the required JSON.
- Crew AI decision agent returns the correct BUY/SELL/HOLD payload.
- One end-to-end integration test passes under performance budget.

Agents Overview:

InferenceAgent
DecisionAgent
ExplainAgent

**Detailed Agent Descriptions:**

InferenceAgent

Loads CNN+Attention+LSTM and XGBoost models.

Processes a new price window and emits { cnn_lstm_score, xgb_prob }.

DecisionAgent

Reads the JSON, applies:

xgb_prob > 0.6 → BUY

xgb_prob < 0.4 → SELL

otherwise → HOLD

Outputs {"recommendation":"…"}.

ExplainAgent

Takes the same JSON plus XGBoost feature importances.

Generates a brief natural-language rationale explaining the recommendation.

**Tasks (with person-hours):**

- [ ] XGB.1 Data Ingestion & Windowing
- [ ] XGB.1.1 Fetch OHLCV for target tickers via yfinance (4 ph)
- [ ] XGB.1.2 Impute or drop missing data; align timestamps (4 ph)
- [ ] XGB.1.3 Normalize features (e.g., z-score) (4 ph)
- [ ] XGB.1.4 Slice into overlapping windows of length T; create train/val/test splits (8 ph)

- [ ] XGB.2 CNN-Attention-LSTM Pretraining
- [ ] XGB.2.1 Implement 1D-CNN encoder (PyTorch or Keras) (8 ph)
- [ ] XGB.2.2 Add self-attention layer over time steps (6 ph)
- [ ] XGB.2.3 Stack an LSTM decoder with regression/classification head (6 ph)
- [ ] XGB.2.4 Write training loop with loss, optimizer, checkpointing, and early stopping (10 ph)
- [ ] XGB.2.5 Log training/validation metrics and save best model (4 ph)

- [ ] XGB.3 Embedding Extraction
- [ ] XGB.3.1 Freeze the trained CNN+Attention+LSTM backbone (4 ph)
- [ ] XGB.3.2 Forward all windows through the network, extracting final hidden-state vectors (4 ph)
- [ ] XGB.3.3 Save embeddings and corresponding labels to disk (e.g., CSV or HDF5) (4 ph)

- [ ] XGB.4 XGBoost Fine-tuning
- [ ] XGB.4.1 Load embeddings + labels into a pandas DataFrame (2 ph)
- [ ] XGB.4.2 Perform train/validation split (2 ph)
- [ ] XGB.4.3 Grid-search n_estimators, max_depth, learning_rate (8 ph)
- [ ] XGB.4.4 Train the best XGBRegressor or XGBClassifier; evaluate and save the model (4 ph)

- [ ] XGB.5 Inference Wrapper
- [ ] XGB.5.1 Write a Python script that loads both the deep model and XGBoost (4 ph)
- [ ] XGB.5.2 Given the latest window, compute cnn_lstm_score and xgb_prob (4 ph)
- [ ] XGB.5.3 Emit JSON with date, scores, and probabilities (2 ph)

- [ ] XGB.6 Crew AI Agent Development
- [ ] XGB.6.1 DecisionAgent: Implement prompt logic to map xgb_prob to BUY/SELL/HOLD (4 ph)
- [ ] XGB.6.2 ExplainAgent: Build an agent that ingests the same JSON plus feature importances and outputs a concise rationale (4 ph)
- [ ] XGB.6.3 End-to-end test chaining InferenceAgent → DecisionAgent → ExplainAgent (6 ph)
- [ ] 
- [ ] XGB.7 Integration Testing & Performance
- [ ] XGB.7.1 Write integration tests covering data fetch → model inference → Crew AI decision (6 ph)
- [ ] XGB.7.2 Benchmark end-to-end latency; optimize any slow steps to meet the ≤ 5 s target (6 ph)

- [ ] XGB.8 Backtesting & Evaluation
- [ ] XGB.8.1 Prepare historical test sets and data conversion (4 ph)
- [ ] XGB.8.2 Implement backtesting harness to simulate the full inference + decision pipeline over history (6 ph)
- [ ] XGB.8.3 Calculate performance metrics (e.g., accuracy, return, drawdown) and analysis (4 ph)
- [ ] XGB.8.4 Generate a backtesting report and visualizations (4 ph)
